File: dotnetcore/Dependencies/FileBasedDependency/FileDependency/FileDependency.cs
// ===============================================================================
// Alachisoft (R) NCache Sample Code.
// ===============================================================================
// Copyright © Alachisoft.  All rights reserved.
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
// OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
// LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE.
// ===============================================================================

using System;
using System.Threading;
using Alachisoft.NCache.Runtime;
using Alachisoft.NCache.Sample.Data;
using System.Configuration;
using Alachisoft.NCache.Client;
using Alachisoft.NCache.Runtime.Dependencies;
using System.IO;

namespace Alachisoft.NCache.Samples
{
    /// <summary>
    /// This class implements file based dependency
    /// </summary>
	public class FileBasedDependency
	{
        private static ICache _cache;

        /// <summary>
        /// Executing this method will perform all the operations to use file based dependency
        /// </summary>
		public static void Run()
		{
			// Initialize cache
            InitializeCache();

            // Add file based dependency 
            AddFileBasedDependency();

            // Dispose the cache once done
			_cache.Dispose();
		}

        /// <summary>
        /// This method initializes the cache
        /// </summary>
        private static void InitializeCache()
        {
            string cache = ConfigurationManager.AppSettings["CacheID"];

            if (String.IsNullOrEmpty(cache))
            {
                Console.WriteLine("The CacheID cannot be null or empty.");
                return;
            }

            // Initialize an instance of the cache to begin performing operations:
            _cache =CacheManager.GetCache(cache);

            // Print output on console
            Console.WriteLine(string.Format("\nCache '{0}' is initialized.", cache));
        }

        /// <summary>
        /// This method adds file based dependency 
        /// </summary>
        private static void AddFileBasedDependency()
        {
            string basePath = Directory.GetCurrentDirectory().Split(new string[] { "\\FileDependency\\bin" }, StringSplitOptions.None)[0];
            string dependencyfile = basePath + "\\DependencyFile\\foobar.txt";
            Console.WriteLine(basePath);
            // Generate a new instance of product 
            Product product = new Product { Id = 52, Name = "Filo Mix", Category = "Grains/Cereals", UnitPrice = 46 };

            // Adding item dependent on file 
            CacheItem cacheItem = new CacheItem(product);
            cacheItem.Dependency = new FileDependency(dependencyfile);
            _cache.Add("Product:"+product.Id, cacheItem);

            Console.WriteLine("\nItem 'Product:52' added to cache with file dependency. ");

            // To verify that the dependency is working, uncomment the following code.
            // Any change in the file will cause invalidation of cache item and thus the item will be removed from cache.
            // Following code modifies the file and then verifies the existence of item in cache.

            //// Modify file programmatically
            ModifyDependencyFile(dependencyfile);
            Console.ReadKey();
            ////... and then check for its existence
            object item = _cache.Get<object>("Product:" + product.Id);
            if (item == null)
            {
                Console.WriteLine("Item has been removed due to file dependency.");
            }
            else
            {
                Console.WriteLine("File based dependency did not work. Dependency file located at " + dependencyfile + " might be missing or file not changed within the given interval.");
            }

            // Remove sample data from Cache to 
            _cache.Remove("Product:"+product.Id);
        }

        /// <summary>
        /// This method modifies dependency file programmatically
        /// </summary>
        /// <param name="path"> Absolute path of dependency file </param>
        private static void ModifyDependencyFile(string path)
        {
            using (System.IO.StreamWriter streamWriter = new System.IO.StreamWriter(path, true))
            {
                streamWriter.WriteLine(string.Format("\n{0}\tFile is modifed. ", DateTime.Now));
            }

            Console.WriteLine(string.Format("File '{0}' is modified programmatically.", path));
        }
	}
}

File: python/Dependencies/FileBasedDependency/FileDependency/FileDependency.py
import os
import time
from datetime import datetime
from ncache.client.CacheManager import CacheManager
from ncache.client.CacheItem import CacheItem
from ncache.runtime.dependencies.FileDependency import FileDependency
from SampleData.Product import Product


class FileBasedDependency:
    cache = None

    @staticmethod
    def run():
        # Initialize cache
        FileBasedDependency.initialize_cache()
        # Add a single key and modify file to see if dependent is removed
        FileBasedDependency.add_file_based_dependency()
        # Dispose the cache once done
        FileBasedDependency.cache.close()

    @staticmethod
    def initialize_cache():
        cache_id = "demoCache"
        FileBasedDependency.cache = CacheManager.get_cache(cache_id)
        print(f"\nCache '{cache_id}' is initialized.")

    @staticmethod
    def add_file_based_dependency():
        dependency_file = os.path.join(os.path.abspath(os.path.join(os.getcwd(), os.pardir)), "DependencyFile", "foobar.txt")
        product = Product(id="1001", name="Tea", unit_price=50, units_available=20)
        cache_item = CacheItem(product)
        cache_item.set_dependency(FileDependency(dependency_file))
        FileBasedDependency.cache.add(f"Product:{product.id}", cache_item)
        print(f"\nItem 'Product:{product.id}' added to cache with file dependency.")

        # Any change in the file will cause invalidation of the cache item.
        # Following code modifies the file and then verifies the existence of the item in the cache.
        time.sleep(1)
        # Modify file programmatically
        FileBasedDependency.modify_dependency_file(dependency_file)
        time.sleep(30)

        item = FileBasedDependency.cache.get(f"Product:{product.id}", Product)
        if item is None:
            print("\nItem has been removed due to file dependency.")
        else:
            print("\nFile-based dependency did not work. Dependency file located at {dependency_file} might be missing or file not changed within the given interval.")

    @staticmethod
    def modify_dependency_file(path):
        with open(path, "a") as file:
            file.write(f"\n{datetime.now()}\tFile is modified.")

        print(f"\nFile '{path}' is modified programmatically.")


FileBasedDependency.run()
